src/selectors_vm/ast.rs

```diff
@@ -1,86 +1,120 @@
+    /// A first child with a step of 0 and an offset of 1
+    pub fn first() -> Self {
+        Self::new(0, 1)
+    }
+
+    pub fn new(step: i32, offset: i32) -> Self {
+        Self { step, offset }
+    }
+
+    pub fn has_index(self, index: i32) -> bool {
+        let Self { offset, step } = self;
+        let offsetted = index - offset;
+        if step == 0 {
+            offsetted == 0
+        } else if (offsetted < 0 && step > 0) || (offsetted > 0 && step < 0) {
+            false
+        } else {
+            (offsetted % step) == 0
+        }
+    }
 }
 
 #[derive(PartialEq, Eq, Debug)]
-pub enum NonAttributeExpr {
+pub enum OnTagNameExpr {
     ExplicitAny,
     Unmatchable,
     LocalName(String),
+    NthChild(NthChild),
+    NthOfType(NthChild),
 }
 
-enum SimpleExpr {
-    NonAttributeExpr(NonAttributeExpr),
-    AttributeExpr(AttributeExpr),
+#[derive(Eq, PartialEq)]
+pub struct AttributeExpr {
+    pub name: String,
+    pub value: String,
+    pub case_sensitivity: ParsedCaseSensitivity,
+    pub operator: AttrSelectorOperator,
 }
 
-impl SimpleExpr {
+impl AttributeExpr {
     #[inline]
-    fn attr_expr_for_operator(
-        operator: AttrSelectorOperator,
-        name: &str,
-        value: &str,
+    pub fn new(
+        name: String,
+        value: String,
         case_sensitivity: ParsedCaseSensitivity,
-    ) -> AttributeExpr {
-        use AttrSelectorOperator::*;
-
-        let operand = AttributeExprOperand {
-            name: name.to_owned(),
-            value: value.to_owned(),
-            case_sensitivity,
-        };
+        operator: AttrSelectorOperator,
+    ) -> Self {
+        Self { name, value, operator, case_sensitivity }
+    }
+}
 
-        match operator {
-            DashMatch => AttributeExpr::AttributeDashMatch(operand),
-            Equal => AttributeExpr::AttributeEqual(operand),
-            Includes => AttributeExpr::AttributeIncludes(operand),
-            Prefix => AttributeExpr::AttributePrefix(operand),
-            Substring => AttributeExpr::AttributeSubstring(operand),
-            Suffix => AttributeExpr::AttributeSuffix(operand),
-        }
+impl Debug for AttributeExpr {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        f.debug_struct("AttributeExpr")
+            .field("name", &self.name)
+            .field("value", &self.value)
+            .field("case_sensitivity", &self.case_sensitivity)
+            .field("operator", match self.operator {
+                AttrSelectorOperator::Equal     => &"AttrSelectorOperator::Equal",
+                AttrSelectorOperator::Includes  => &"AttrSelectorOperator::Includes",
+                AttrSelectorOperator::DashMatch => &"AttrSelectorOperator::DashMatch",
+                AttrSelectorOperator::Prefix    => &"AttrSelectorOperator::Prefix",
+                AttrSelectorOperator::Substring => &"AttrSelectorOperator::Substring",
+                AttrSelectorOperator::Suffix    => &"AttrSelectorOperator::Suffix",
```

Comment:
This doesn't seem to be formatter with `cargo fmt`

Reply:
`cargo fmt` can't format it so I added the extra space to line up everything.